### ISSUE TITLE
Performance regression in delayed pipeline in 0.14

### ISSUE THREAD

**Issue by pwaller** [dask/dask#2285]
In summary, my dask delayed pipeline is ~3-4x slower after upgrading from 0.11.1 to 0.14.1.

My pipeline looks like this, where files is a list of filenames and `load` constructs a pandas dataframe by loading msgpack serialised data from a file. I have thousands of files so this regression is a significant runtime hit for me - but the issue is shown here on just the first 32. I'm running on a 16 core machine.

```python
dask.set_options(get=dask.multiprocessing.get)
dfs = [delayed(load)(fn) for fn in files[:32]]
df = dd.from_delayed(dfs)
df.compute()
```

I have bisected this issue to 259b51f50a26b4839ea117f1d2c48dab1170dcb7 (#1919).

Below I show profiles collected before and after this commit.

/cc @eriknw @mrocklin (PR author, merger respectively)

## Before

Hover shows that the task is called `load(*)`.

![before](https://cloud.githubusercontent.com/assets/438648/25613717/2631f5aa-2f27-11e7-9889-ad27cd52c59b.png)

## After

Hover shows that the dark blue task is called `load(*)` and the other turquoise tasks are called `_`.

![after](https://cloud.githubusercontent.com/assets/438648/25613718/26331188-2f27-11e7-8a84-74930d44aa70.png)

**Comment by jcrist** [comment]
Ok, so the issue here is that you now have twice as many tasks, with one task being an alias to another task. This isn't optimized away, leading to twice as much communication due to how the multiprocessing scheduler works. I see two fixes:

- Use the distributed scheduler instead. The distributed scheduler can be much smarter about how data moves around, whereas the multiprocessing scheduler has to bring intermediates back to the main process after every task. Please read more here: http://dask.pydata.org/en/latest/scheduler-choice.html#distributed-scheduler-on-a-single-machine.

- Add a `dealias` step to the optimizations performed by the multiprocessing scheduler. This would make this workload just as efficient as before using the multiprocessing scheduler.

I recommend 1 over 2.

**Comment by mrocklin** [comment]
+1 on using the distributed scheduler if possible.  The multiprocessing scheduler is not the greatest.  

You can start a local distributed cluster with the following lines:

```python
from dask.distributed import Client
client = Client()
```

Unfortunately your existing diagnostic code will start working (the diagnostics are married to the particular scheduler.)  However fortunately you should be able to use the web diagnostic page at http://localhost:8787/status

**Comment by mrocklin** [comment]
Also, thank you for going through the trouble of bisecting to identify the commit.  That is incredibly helpful.

**Comment by pwaller** [comment]
Thanks for the replies.

> the issue here is that you now have twice as many tasks

I would love to understand **why** that is. Is the code fundamentally incorrect or unsupported, or is it a bug? I arrived at my code by straightforwardly copying the documentation. I didn't see anything that would suggest I should expect twice as many tasks and my mental model currently lacks a mechanism for explaining this. Any help improving my understanding appreciated.

Is the fundamental recommendation here to stop using the multiprocessing scheduler?

**Comment by jcrist** [comment]
The alias is no longer fused away due to the change in how `fuse` (an optimization internal to `dask.dataframe`) works. This is not a bug in your code or our code, just a change in how things work internally. More information on why this change happened can be seen in the issue for the PR: https://github.com/dask/dask/issues/1911. 

> Is the fundamental recommendation here to stop using the multiprocessing scheduler?

Yes. For problems that are basically just a `map` operation it can be efficient, but for anything more complicated the distributed scheduler will be more efficient. It also has better diagnostics: http://distributed.readthedocs.io/en/latest/web.html.

**Comment by jcrist** [comment]
One option for an immediate fix is to use some of the tuning settings for fuse to generate a better graph. This is a bit more low-level, but can result in the same graph you had before:

```python
In [1]: import dask.dataframe as dd

In [2]: import dask

In [3]: @dask.delayed
   ...: def load(x):
   ...:     pass
   ...:

In [4]: parts = [load(i) for i in range(4)]

In [5]: ddf = dd.from_delayed(parts, meta=[('x', int), ('y', int)])

In [6]: ddf.visualize(filename='before.png', optimize_graph=True)
Out[6]: <IPython.core.display.Image object>

In [7]: with dask.set_options(fuse_ave_width=1, fuse_rename_keys=False):
   ...:     ddf.visualize(filename='after.png', optimize_graph=True)
```

### Before:
![before](https://cloud.githubusercontent.com/assets/2783717/25625576/7fdf4742-2f22-11e7-9e87-7d0d43b3cf9f.png)


### After:
![after](https://cloud.githubusercontent.com/assets/2783717/25625584/864fbf26-2f22-11e7-8fa7-fe59c5287089.png)

For your use, you'd put the `dask.set_options(...)` contextmanager around your call to `compute`.

**Comment by jcrist** [comment]
@pwaller is this issue resolved? To summarize, the recommendation was:

- Use the distributed scheduler locally, which in most cases will perform better than the multiprocessing scheduler, and is just as easy to setup.
- Or override the `fuse` settings using `dask.set_options`. This will fix your current performance issue, but if you ever try to do something more complicated than a simple map operation you'll see performance issues again (which would be solved by using the distributed scheduler instead).

**Comment by pwaller** [comment]
Hi, Apologies for the delay, I was on holiday :palm_tree:. Last I recall, I was using an unsatisfying workaround. I'll give your suggestions a go when I next need to do this.

It looks like you guys thoroughly understand the problem, and I'm happy my particular problem is resolved, so I'm happy to close the issue for now.

**However**: as a beginner starting to use dask it was a bit of an alarming issue`*` to encounter with no obvious explanation, and if I were doing all this again 6 months from now I would likely run headlong into this. So maybe you would like to keep and somehow solve the issue, even if it is a matter of updating the documentation or adding deprecation notices to tell people not to use the process scheduler? It would have saved me some suffering, at least.

I'll leave the close button in your court.

`*`  (I originally wrote "bug" here, but you may disagree about whether the behaviour change between versions and the slow performance for "the obvious code to write for a beginner" is a bug or not)

**Comment by pwaller** [comment]
Also to note: if I had not spotted that this was a regression (e.g, if I had just started off using today's version), I would likely have given up on dask for being too slow.

My natural reading of the documentation looking for how to do parallelism (and then stop, because that's the code I need to get the work done) leads me to [embarassingly-parallel-computation](http://dask.pydata.org/en/latest/use-cases.html#embarrassingly-parallel-computation), where `multiprocessing.get` sounds a heck of a lot simpler and more reliable than doing `Cluster`, especially if you don't actually have a cluster of machines to play with for the task at hand.

And since I'm just extruding my thoughts here - the documentation read linearly from the beginning in general does not make it very clear how to achieve parallelism. That configuration knob really belongs somewhere at the beginning of the documentation. I was puzzled for a while why the first few dask programs I wrote weren't parallel at all.

(I'm casting myself back to before I had read the documentation more thoroughly to give you a hint at how I arrived at this position).

**Comment by pwaller** [comment]
I have just tried replacing my `multiprocessing.get` with:

```python3
from dask.distributed import Client
client = Client()
dask.set_options(get=client.get)
```

And unfortunately I get timings comparable with the performance regression.

Also, my `localhost:8787/status`  just hangs. I get an `(52) Empty reply from server` when I try to curl it, after a long period of time. `localhost:8787/` does load, offering me links to various things.

**Comment by pwaller** [comment]
Edit to the above: Accidentally, I was not comparing like with like. I still have the problem of the distributed web interface hanging and then giving an empty reply.

**Comment by jcrist** [comment]
> I still have the problem of the distributed web interface hanging and then giving an empty reply.

This might be an issue with a release of tornado (4.5.0) that broke bokeh. If this is the case, upgrading to 4.5.1 should fix this (downgrading would work too).

**Comment by pwaller** [comment]
I'm on tornado 4.5.1.

**Comment by jcrist** [comment]
Not sure then. Pinging @mrocklin for more knowledge about the bokeh diagnostics.

**Comment by canavandl** [comment]
I don't think tornado 4.5.1 is compatible with bokeh 0.12.5 either. You'll either have to downgrade tornado or install a dev release of bokeh from the 'bokeh/label/dev' channel.

**Comment by mrocklin** [comment]
What is the status on this?  Can this be closed?

**Comment by pwaller** [comment]
Essentially, I'm happy if you want to close this issue.

**TL;DR of the below**: I almost gave up on dask. Please read the actionable stuff at the bottom and consider whether it's worth fixing or having an open issue. Feel free to close this issue.

I hope my 2¢ are useful here, really want to see you guys succeed in the single-beefy-machine use case :).

---

I was unable to get bokeh working even by downgrading tornado. But I don't have a strong need to get that working (`</bokeh sub-thread>`).

I did get a workflow with the `distributed.Client` which works OK and gives me reasonable performance. I recall that I found the performance worse without the `fuse_` optimizations switched on - so I don't understand why they aren't default or when I would want to switch them off. Here's what I'm using:

```
from dask.distributed import Client
client = Client(ip="1.2.3.4")
dask.set_options(get=client.get, fuse_ave_width=1, fuse_rename_keys=False)
```

I don't know if you guys want to open an issue up for the bokeh problem or my [comments on what it's like as someone new to dask](https://github.com/dask/dask/issues/2285#issuecomment-301337670) -in my view, it could affect the uptake of dask and almost caused me to give up on it.

I saw `distributed.Client` and assumed it was irrelevant to me because I'm on a single machine with many cores. My expectation was that `distributed` could only make the performance worse and the configuration and setup more complicated. I'm looking for quick and cheap options and the word "distributed" immediately implies the opposite. The documentation doesn't correct this misconception.

The front of your docs also contains this line:

> "Dask currently has two main schedulers, one for single machine processing using threads or processes, and one for distributed memory clusters."

This says to me "ignore the distributed one and use multiprocessing", because I only have one beefy machine.

# The actionable things here are to:

* modify your documentation to remove or make less prominent the parts about `multiprocessing.get`, or otherwise add appropriate caveats about the negative performance implications.
* make the "embarrassingly parallel getting started" more prominent earlier in the documentation
  * currently it's buried a long way in and hard to find at all from the front page, even when I know specifically what I'm looking for.
  * [the front page of the docs](http://dask.pydata.org/en/latest/) talks a lot about the architecture of dask and that it is parallel, but neglects to make it obvious to a beginner how to execute code in parallel. (e.g, you mention schedulers, but none of the examples under `Familiar user interface` mention anything to do with scheduling). The word `scheduling` is a distant cousin of the word `parallel`, it's not obvious at first glance they are related.
  * To see what I mean here, just go to the docs and do CTRL-F "parallel", and try and find some code you should use. I cannot find any, after some looking.
* make it clear that `distributed.Client.get`:
    * doesn't require a cluster
    * doesn't impose a performance penalty when you're not using a cluster
    * does not require any additional configuration compared with multiprocessing.get
    * is easy to use

**Comment by mrocklin** [comment]
Thanks for the thoughful comments @pwaller 

The Bokeh issues should be resolved in their next release, which is sometime next week (I think).  I am as frustrated as you are here and so really glad to see this resolved.

Optimal values for optimization vary between applications.  There are several prominent applications that have the reverse experience from yours here.  We've tried to choose values that are decent in most cases.  It sounds like your case was not in this set.  I'm not sure how to resolve this.

Making multiprocessing.get less prominent sounds like a reasonable suggestion.   Where did you run into this term, [this page](http://dask.pydata.org/en/latest/scheduler-overview.html)?

> the front page of the docs talks a lot about the architecture of dask and that it is parallel, but neglects to make it obvious to a beginner how to execute code in parallel. (e.g, you mention schedulers, but none of the examples under Familiar user interface mention anything to do with scheduling). The word scheduling is a distant cousin of the word parallel, it's not obvious at first glance they are related.

I'm not sure I understand the concern here.  All of these examples in this section run in parallel.  Should we make it more clear that all of these examples run in parallel?

Yes, I agree that we need to lower the mental barrier to using dask.distributed.  We've tried to make this prominent in the scheduling docs but should probably do more.

Thank you again for your feedback.  More detailed suggestions, or even pull requests, would be welcome.

**Comment by pwaller** [comment]
I hope this doesn't come off as ranty - it's not intended that way - and apologies if it does - I'm just trying to provide my perspective of dask when I was beginning to learn to use it, and explain my confusion and difficulties with being pointed at the `distributed` scheduler and `multiprocessing` not being the thing to use.

> I'm not sure I understand the concern here. All of these examples in this section run in parallel. 

~Only if you specify a parallel scheduler, no? As far as I can tell, that is not apparent from the documentation. Forgive me (and tell me!) if I'm wrong here, I don't have data in formats where I can just copy and paste the examples (I have to use delayed with a load function).~ Edit: I have finally figured one problem out: that the default threaded scheduler doesn't work in my use case where I call functions which don't release the GIL. Please read the ~strikeout~ text to see the confusion that was held until this moment.

If you read the docs linearly, this is the first thing you hit which is promising which explains how to actually do parallel computation (in general, for functions that don't release the GIL), that you need to specify the `get` function:

http://dask.pydata.org/en/latest/use-cases.html#embarrassingly-parallel-computation

If you instead go to the examples page:

http://dask.pydata.org/en/latest/examples-tutorials.html

The only example is "[Build Parallel Algorithms for Pandas](https://gist.github.com/mrocklin/d009e5d4a1f49ecdb433107f3d72c7f3#file-pygotham-dataframes-ipynb)", which again, doesn't teach you anything about schedulers. It also won't do parallel computation if your computation involves mostly functions which hold the GIL.

As soon as you start thinking "Maybe I should look into this schedulers thing", the first links you hit in the side bar are http://dask.pydata.org/en/latest/distributed.html; that sends you off to learn about the [nitty gritty](https://distributed.readthedocs.io/en/latest/) of the distributed scheduler. Again, this is problematic because those docs are a massive wall of text and features which obscure the fact it is appropriate to use in my use case (single beefy machine). It's also problematic because when I reached this point, I was wanting to learn about `dask`, but the first thing I see in the [Scheduler Overview](http://dask.pydata.org/en/latest/scheduler-overview.html) is:

> distributed.Client.get: a distributed scheduler for executing graphs on multiple machines. This lives in the external distributed project.

Which again, is easy to interpret as a hint that I don't want to use it because it's this whole other dependency and project that I have to go learn about, fetch and make work. One that doesn't sound directly relevant to the task I'm trying to solve (use all 64 cores in my machine).

# Summary of my critique

The front-facing docs (the landing page, the examples and use cases) expose too much internal detail while not providing an obvious direct route to parallel computation (if you happen to hold the GIL). The current "obvious route" is to use `multiprocessing` and even if one were curious enough to discover the `distributed` scheduler, it very much appears from a distance to be this heavy thing unrelated to many-core-single-machine GIL-holding computation.

All of my first steps with `dask` involved tasks (importing the data in particular) which happened to hold the GIL, so I struggled to get a parallel speedup for anything until I found `multiprocessing`, which was then lost when I upgraded dask.

**Comment by pwaller** [comment]
Just to soften my critique a little: I am grateful for dask and the documentation. I think it's a good project, it just needs subtle tweaks to ease the learning curve - in particular for GIL using code.

**Comment by mrocklin** [comment]
No need to soften the critique @pwaller .  They're valid points and you've expressed them clearly.  

In general I agree that the path through the docs could be made cleaner.  Generally I've found this challenging because there are so many different paths that people should take.  We often get the complaint that "you should really put X front and center on the docs, it's what really matters" for a surprisingly broad set of items X.  The fact that you are running GIL-bound computations may actually not be that typical among dask users (or it may be, I don't really know).  Providing a path through the documentation that is generic enough for all communities is surprisingly difficult.

That being said there are clearly things that could be done to improve the situation.  If you really want to soften the critique you could follow it up with concrete suggestions on exactly what to, which sections to move where, etc..

**Comment by pwaller** [comment]
I think the best I can do at the moment is the list of actionable things I already submitted [at the bottom of this comment](https://github.com/dask/dask/issues/2285#issuecomment-305799980). I really want to help out more, but I'm afraid I have other priorities. I think to make any more coherent impact here (e.g, to actually submit PRs for the docs) I'd need to work closely with whoever wrote/maintained the docs, which I don't have time to do. Hopefully whoever that is can mix my suggestions from this thread into their own ideas.

This thread is already very long, so I suggest closing it for now. If there is anything in here which is important enough for you to track in a github issue, I guess it would be better to start a separate issue?

All the best!

**Comment by mrocklin** [comment]
Thanks for the comments!
